src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp

```diff
@@ -490,6 +495,105 @@ void RenderingManager::slot_camera_change_end()
 {
 }
 
+void RenderingManager::slot_material_dropped(const foundation::Vector2d& drop_pos, const std::string& material_name)
+{
+    if (!m_project->has_trace_context())
+    {
+        RENDERER_LOG_INFO("the scene must be rendering or must have been rendered at least once for drag&drop to be available.");
+        return;
+    }
+
+    DroppedMaterialMenu* menu = new DroppedMaterialMenu(drop_pos, material_name);
+
+    connect(menu,
+        SIGNAL(signal_apply_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)),
+        SLOT(slot_change_material(
+            const foundation::Vector2d&,
+            const std::string&,
+            renderer::ObjectInstance::Side)));
+
+    menu->exec(QCursor::pos());
+}
+
+namespace
+{
+    struct MaterialSlot
+    {
+        std::string name;
+        ObjectInstance::Side side;
+    };
+
+    class ApplyDroppedMaterial
+        : public RenderingManager::IScheduledAction
+    {
+    public:
+        explicit ApplyDroppedMaterial(
+            ObjectInstance& instance,
+            const std::vector<MaterialSlot>& material_slots,
+            const std::string& material_name)
+            : m_instance(instance),
+            m_material_slots(material_slots),
+            m_material_name(material_name)
+        {
+        }
+
+        void operator()(Project& project) override
+        {
+            for (const_each<std::vector<MaterialSlot>> slot = m_material_slots; slot; ++slot)
```

Comment:
Use modern range-based for loop. `foundation::const_each<>` is deprecated, we're in the process of converting them all to their modern counterparts (you can help too!).